### PR TITLE
spirv-llvm-translator: 8.0.1 -> 12.0.0

### PR DESCRIPTION
**Pull request by vobloeb**
llvm 8 is too old, updated to llvm 12
also enable test suite

### CODE REVIEW

File: pkgs/development/compilers/spirv-llvm-translator/default.nix [hunk 1]
```diff
@@ -1,33 +1,36 @@
...
+                 "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
+               ];
+  # also make `llvm-spirv`
+  postBuild = "make llvm-spirv";
+
+  doCheck = true;
 
-  # FIXME: CMake tries to run "/llvm-lit" which of course doesn't exist
-  doCheck = false;
+  postInstall = "install -Dm755 tools/llvm-spirv/llvm-spirv -t ${placeholder "out"}/bin";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/spirv-llvm-translator/default.nix:33]:
```suggestion
  postInstall = ''
    install -Dm755 tools/llvm-spirv/llvm-spirv -t ${placeholder "out"}/bin
  '';
```

File: pkgs/development/compilers/spirv-llvm-translator/default.nix [hunk 2]
```diff
@@ -1,33 +1,36 @@
...
 
-  buildInputs = [ llvm_8 ];
+  buildInputs = [ llvmPackages_12.llvm ];
 
   checkInputs = [ lit ];
 
-  cmakeFlags = [ "-DLLVM_INCLUDE_TESTS=ON" ];
+  cmakeFlags = [ "-DLLVM_INCLUDE_TESTS=ON"
+                 "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
+               ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/spirv-llvm-translator/default.nix:27]:
```suggestion
  cmakeFlags = [
    "-DLLVM_INCLUDE_TESTS=ON"
    "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
  ];
```

File: pkgs/development/compilers/spirv-llvm-translator/default.nix [hunk 3]
```diff
@@ -1,33 +1,39 @@
...
 
   checkInputs = [ lit ];
 
-  cmakeFlags = [ "-DLLVM_INCLUDE_TESTS=ON" ];
+  cmakeFlags = [
+    "-DLLVM_INCLUDE_TESTS=ON"
+    "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
+  ];
+  # also make `llvm-spirv`
+  postBuild = "make llvm-spirv";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/spirv-llvm-translator/default.nix:30]:
```suggestion
  postBuild = ''
    make llvm-spirv
  '';
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 130127` at 46fa406d run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>Literate</li>
    <li>dtools</li>
    <li>gtkd</li>
    <li>intel-compute-runtime</li>
    <li>onedrive</li>
    <li>rund</li>
    <li>tilix</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/ff555abf71d62765884218d0fa39f461">intel-graphics-compiler</a></li>
    <li><a href="https://gist.github.com/5606d9815820257bd7742ba15c21f834">opencl-clang</a></li>
    <li><a href="https://gist.github.com/62409c9369768fa20b6d042dd5854308">spirv-llvm-translator</a></li>
  </ul>
</details>
<details>
  <summary>4 packages skipped due to time constraints:</summary>
  <ul>
    <li>lldb (lldb_12 ,llvmPackages_12.lldb ,llvmPackages_latest.lldb)</li>
    <li>lldb_10 (llvmPackages_10.lldb)</li>
    <li>lldb_11 (llvmPackages_11.lldb)</li>
    <li>vscode-extensions.vadimcn.vscode-lldb</li>
  </ul>
</details>
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>lit</li>
    <li>lldb_9 (llvmPackages_9.lldb)</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/development/compilers/spirv-llvm-translator/default.nix [hunk 4]
```diff
@@ -1,33 +1,41 @@
...
+  buildInputs = [ llvmPackages_12.llvm ];
 
   checkInputs = [ lit ];
 
-  cmakeFlags = [ "-DLLVM_INCLUDE_TESTS=ON" ];
+  cmakeFlags = [
+    "-DLLVM_INCLUDE_TESTS=ON"
+    "-DLLVM_EXTERNAL_LIT=${lit}/bin/lit"
+  ];
+  # also make `llvm-spirv`
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/spirv-llvm-translator/default.nix:29]:
```suggestion

```

File: pkgs/development/compilers/spirv-llvm-translator/default.nix [hunk 5]
```diff
@@ -1,33 +1,41 @@
...
+  postBuild = ''
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/spirv-llvm-translator/default.nix:30]:
```suggestion
  ];

  postBuild = ''
```

**Comment by SuperSandro2000** [comment]
Please squash the spirv-llvm-translator commits.

**Comment by risicle** [comment]
Builds fails on macos 10.15:

```
builder for '/nix/store/qypy7p9w5jvgpjf9yg1rack9b4hmrfws-spirv-llvm-translator-12.0.0.drv' failed with exit code 1; last 10 log lines:
  --------------------------
  |diff --git a/test/transcoding/enqueue_kernel.ll b/test/transcoding/enqueue_kernel.ll
  |index 1f0b360..761043e 100644
  |--- a/test/transcoding/enqueue_kernel.ll
  |+++ b/test/transcoding/enqueue_kernel.ll
  --------------------------
  File to patch:
  Skip this patch? [y]
  Skipping patch.
  9 out of 9 hunks ignored
```

**Comment by risicle** [comment]
For the actual package behind the attr `spirv-llvm-translator`, build results in

```
[100%] Linking CXX executable llvm-spirv
Undefined symbols for architecture x86_64:
  "operator new[](unsigned long)", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in llvm-spirv.cpp.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::setbuf(char*, long) in llvm-spirv.cpp.o
  "operator new(unsigned long)", referenced from:
      _main in llvm-spirv.cpp.o
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::handleOccurrence(unsigned int, llvm::StringRef, llvm::StringRef) in llvm-spirv.cpp.o
      void std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::__push_back_slow_path<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in llvm-spirv.cpp.o
      std::__1::__function::__func<llvm::ExitOnError::ExitOnError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int)::'lambda'(llvm::Error const&), std::__1::allocator<llvm::ExitOnError::ExitOnError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int)::'lambda'(llvm::Error const&)>, int (llvm::Error const&)>::__clone() const in llvm-spirv.cpp.o
      std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<unsigned int, unsigned long long>, void*>*>, bool> std::__1::__hash_table<std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::__unordered_map_hasher<unsigned int, std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::hash<unsigned int>, std::__1::equal_to<unsigned int>, true>, std::__1::__unordered_map_equal<unsigned int, std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::equal_to<unsigned int>, std::__1::hash<unsigned int>, true>, std::__1::allocator<std::__1::__hash_value_type<unsigned int, unsigned long long> > >::__emplace_unique_key_args<unsigned int, std::__1::piecewise_construct_t const&, std::__1::tuple<unsigned int const&>, std::__1::tuple<> >(unsigned int const&, std::__1::piecewise_construct_t const&, std::__1::tuple<unsigned int const&>&&, std::__1::tuple<>&&) in llvm-spirv.cpp.o
      std::__1::__hash_table<std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::__unordered_map_hasher<unsigned int, std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::hash<unsigned int>, std::__1::equal_to<unsigned int>, true>, std::__1::__unordered_map_equal<unsigned int, std::__1::__hash_value_type<unsigned int, unsigned long long>, std::__1::equal_to<unsigned int>, std::__1::hash<unsigned int>, true>, std::__1::allocator<std::__1::__hash_value_type<unsigned int, unsigned long long> > >::__rehash(unsigned long) in llvm-spirv.cpp.o
      void std::__1::map<SPIRV::ExtensionID, bool, std::__1::less<SPIRV::ExtensionID>, std::__1::allocator<std::__1::pair<SPIRV::ExtensionID const, bool> > >::insert<std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<SPIRV::ExtensionID, bool>, std::__1::__tree_node<std::__1::__value_type<SPIRV::ExtensionID, bool>, void*>*, long> > >(std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<SPIRV::ExtensionID, bool>, std::__1::__tree_node<std::__1::__value_type<SPIRV::ExtensionID, bool>, void*>*, long> >, std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<SPIRV::ExtensionID, bool>, std::__1::__tree_node<std::__1::__value_type<SPIRV::ExtensionID, bool>, void*>*, long> >) in llvm-spirv.cpp.o
      ...
  "operator delete[](void*)", referenced from:
      parseSpecConstOpt(llvm::StringRef, SPIRV::TranslatorOpts&) in llvm-spirv.cpp.o
      std::__1::unique_ptr<llvm::APFloat [], std::__1::default_delete<llvm::APFloat []> >::reset(std::nullptr_t) in llvm-spirv.cpp.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in llvm-spirv.cpp.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in llvm-spirv.cpp.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::setbuf(char*, long) in llvm-spirv.cpp.o
      SPIRV::LLVMToSPIRV::transConstant(llvm::Value*) in libLLVMSPIRVLib.a(SPIRVWriter.cpp.o)
      SPIRV::LLVMToSPIRV::transBuiltinToConstant(llvm::StringRef, llvm::CallInst*) in libLLVMSPIRVLib.a(SPIRVWriter.cpp.o)
      ...
  "operator delete(void*)", referenced from:
      llvm::cl::opt<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~opt() in llvm-spirv.cpp.o
      llvm::ExitOnError::~ExitOnError() in llvm-spirv.cpp.o
      parseSpecConstOpt(llvm::StringRef, SPIRV::TranslatorOpts&) in llvm-spirv.cpp.o
      _main in llvm-spirv.cpp.o
      llvm::cl::opt<SPIRV::VersionNumber, false, llvm::cl::parser<SPIRV::VersionNumber> >::~opt() in llvm-spirv.cpp.o
      llvm::cl::parser<SPIRV::VersionNumber>::~parser() in llvm-spirv.cpp.o
      llvm::cl::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, llvm::cl::parser<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::~list() in llvm-spirv.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
```

on macos 10.15. Unclear what attr the previous error was produced for.

**Comment by jljusten** [comment]
This upgrade could be useful for libclc, which as of version 12 can produce spirv (*.spv) files. Mesa is starting to make use of the spv files.

Refs:
 * https://github.com/llvm/llvm-project/commit/c37145cab12168798a603e22af6b6bf6f606b705
 * https://src.fedoraproject.org/rpms/libclc/c/cde53954336ff6e4f505950766c06d1f0728f646?branch=rawhide
 * https://gitlab.freedesktop.org/mesa/mesa/-/commit/f33b417652ceae711cc34601f51d62beec2d22f1#93dfb713e212614482b153b67399f7b70c5e0086_769_780

**Pull request closed without merging**